### PR TITLE
[HiSparse]Fix DeepSeek V4 HiSparse PD Transfers with Separate Host and Device KV Indices

### LINKED ISSUE #31482: [Bug] DeepSeek-V4 PD HiSparse shows lower accuracy on GPQA and SWE-Bench
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

DeepSeek-V4-Flash-FP8 served with PD disaggregation and decode-side HiSparse produces different logits from an otherwise equivalent non-HiSparse PD deployment. The difference appears while the compressed C4 history is fully device-resident, before host swap-in is required.

The regression has two user-visible forms:

1. GPQA long-reasoning accuracy and convergence degrade.
2. SWE agent runs generate malformed or repetitive DeepSeek-V4 DSML tool-call
   markup, after which the parser returns no usable tool call and the agent
   loops until its step limit.

### Benchmark results

The HiSparse deployment produced:

```text
== gpqa ==
198 examples x 16 repeats  |  11635.6s  |  3375 tok/s  |  39.3M tokens

- pass@1[avg-of-16]  =  81.63% +/- 1.89% (SEM 0.47%)
  pass@16            =  95.45%
  majority@16        =  87.37%
  no_answer          =  4.86%
  stop_rate          =  100.00%
  truncated_rate     =  0.00%
  error_rate         =  0.00%

== gsm8k ==
1319 examples (single-shot)  |  62.2s  |  2281 tok/s  |  142K tokens

- score           =  96.97%
  stop_rate       =  100.00%
  truncated_rate  =  0.00%
  error_rate      =  0.00%
```

The short-generation GSM8K result remains healthy, while regressed GPQA examples are concentrated in long reasoning outputs. HiSparse outputs can also become substantially longer than the non-HiSparse PD baseline:

| Example | Non-HiSparse PD | HiSparse PD |
| --- | ---: | ---: |
| `gpqa-131` | 20,713 tokens | 30,049 tokens |
| `gpqa-47` | 33,176 tokens | 44,268 tokens |
| `gpqa-109` | 9,791 tokens | 29,847 tokens |

### SWE failure mode

In one SWE evaluation of 152 tasks:

- 4 tasks completed normally.
- 144 tasks reached the 300-agent-step limit.
- 4 tasks timed out.

The requests include OpenAI tools, and the `deepseekv4` parser successfully parses many turns into structured calls. Failed turns contain malformed or mixed DSML:

```xml
<tool_raises>
<tool_call>
<invoke name="search">
<parameter name="search_term" string="true">
separability_matrix
</｜DSML｜parameter>
```

Other failed turns enter repetition:

```xml
<｜DSML｜tool_calls>
<invoke name="execute_bash">
<｜DSML｜invoke name="execute_bash">
<｜DSML｜invoke name="execute_bash">
<｜DSML｜invoke name="execute_bash">
...
```

The resulting response can contain:

```text
finish_reason="tool_calls"
tool_calls=[]
```

This is possible because SGLang changes `finish_reason` after detecting a native tool-call marker, before verifying that parsing returned any calls. The agent then records an invalid step, includes the failed output in the next prompt, and repeats the pattern until the 300-step limit.

For one representative task, the non-HiSparse PD deployment completed in 19 steps and produced a 475-character patch. The HiSparse deployment reached 300 steps and produced no patch.

### Source revision

```text
HEAD:
746c67d85fd59958409a7d518dbc0c9618b974ac
Fix DSV4 HiSparse PD state page sizes

Base origin/main:
1b4176cc4674e4f63b2aceb1b39f5747a177ef05
[PD] Fix optimistic prefill inflight-queue hangs on parked/aborted reqs (#31075)
```

### Model and topology

```text
Model: DeepSeek-V4-Flash-FP8
Prefill node: 4 x B200, TP4, EP4
Decode node: 4 x B200, TP4, EP4
PD backend: Mooncake over InfiniBand
HiSparse: decode node only
Context length: 131072
Max running requests: 256
```

### Common environment

```bash
export MODEL_PATH=/path/to/DeepSeek-V4-Flash-FP8
export PREFILL_IP=<prefill-node-ip>
export DECODE_IP=<decode-node-ip>
export PREFILL_API_PORT=30000
export DECODE_API_PORT=30001
export IB_DEVICE=mlx5_3,mlx5_4

export CUDA_VISIBLE_DEVICES=0,1,2,3
export SGLANG_DSV4_FP4_EXPERTS=0
export SGLANG_OPT_FP8_WO_A_GEMM=0
export NVSHMEM_QP_DEPTH=2050
export SGLANG_DEEPEP_NUM_MAX_DISPATCH_TOKENS_PER_RANK=1024

DEEPEP_CONFIG='{"normal_dispatch":{"num_sms":64},"normal_combine":{"num_sms":64}}'
```

### Prefill node

```bash
python3 -m sglang.launch_server \
  --trust-remote-code \
  --model-path "${MODEL_PATH}" \
  --tp 4 \
  --ep 4 \
  --moe-a2a-backend deepep \
  --deepep-config "${DEEPEP_CONFIG}" \
  --tool-call-parser deepseekv4 \
  --reasoning-parser deepseek-v4 \
  --moe-runner-backend deep_gemm \
  --fp8-gemm-backend deep_gemm \
  --chunked-prefill-size 16384 \
  --context-length 131072 \
  --max-running-requests 256 \
  --host "${PREFILL_IP}" \
  --port "${PREFILL_API_PORT}" \
  --mem-fraction-static 0.9 \
  --skip-server-warmup \
  --piecewise-cuda-graph-compiler inductor \
  --allow-auto-truncate \
  --disaggregation-mode prefill \
  --disaggregation-transfer-backend mooncake \
  --disaggregation-ib-device "${IB_DEVICE}" \
  --dist-init-addr "${PREFILL_IP}:30235" \
  --load-balance-method round_robin \
  --enable-metrics \
  --enable-cache-report \
  --swa-full-tokens-ratio 0.1 \
  --tokenizer-worker-num 6 \
  --detokenizer-worker-num 2
```

### Decode node with HiSparse

```bash
HISPARSE_CONFIG='{"top_k":512,"device_buffer_size":2048,"host_to_device_ratio":5}'

python3 -m sglang.launch_server \
  --trust-remote-code \
  --model-path "${MODEL_PATH}" \
  --tp 4 \
  --ep 4 \
  --moe-a2a-backend deepep \
  --deepep-config "${DEEPEP_CONFIG}" \
  --tool-call-parser deepseekv4 \
  --reasoning-parser deepseek-v4 \
  --moe-runner-backend deep_gemm \
  --fp8-gemm-backend deep_gemm \
  --context-length 131072 \
  --max-running-requests 256 \
  --host "${DECODE_IP}" \
  --port "${DECODE_API_PORT}" \
  --mem-fraction-static 0.9 \
  --skip-server-warmup \
  --piecewise-cuda-graph-compiler inductor \
  --allow-auto-truncate \
  --disaggregation-mode decode \
  --disaggregation-transfer-backend mooncake \
  --disaggregation-ib-device "${IB_DEVICE}" \
  --dist-init-addr "${DECODE_IP}:30335" \
  --load-balance-method round_robin \
  --enable-hisparse \
  --hisparse-config "${HISPARSE_CONFIG}" \
  --disable-radix-cache \
  --enable-metrics \
  --enable-cache-report \
  --tokenizer-worker-num 6 \
  --detokenizer-worker-num 2
```

### Environment

Run:

```bash
python3 -m sglang.check_env
```

Output:

```text
Python: 3.12.3 (main, Apr 15 2024, 17:09:49) [GCC 6.3.0 20170516]
CUDA available: True
GPU 0,1,2,3: CF-NG-GBZZ2-O-L
GPU 0,1,2,3 Compute Capability: 10.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.2, V13.2.51
CUDA Driver Version: 580.159.03
PyTorch: 2.11.0+cu130
sglang: 0.0.0
sglang-kernel: 0.4.4+cu130
flashinfer_python: 0.6.14
flashinfer_cubin: Module Not Found
flashinfer_jit_cache: Module Not Found
triton: 3.6.0
transformers: 5.12.1
torchao: 0.17.0
numpy: 2.3.5
aiohttp: 3.14.1
fastapi: 0.139.0
huggingface_hub: 1.23.0
interegular: 0.3.3
modelscope: 1.38.1
orjson: 3.11.9
outlines: 0.1.11
packaging: 26.2
psutil: 7.2.2
pydantic: 2.13.4
python-multipart: 0.0.32
pyzmq: 27.1.0
uvicorn: 0.51.0
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.2.1
openai: 2.6.1
tiktoken: 0.13.0
anthropic: 0.116.0
litellm: Module Not Found
torchcodec: Module Not Found
```

### PR DESCRIPTION
**Pull request by huangtingwei9988**
fix #31482 

## Motivation

This PR fixes corrupted outputs in DeepSeek V4 when using HiSparse with PD disaggregation and pipeline parallelism.
The direct-to-host transfer path previously used host KV page indices for both host and device destination buffers. However, HiSparse allocates host pages and logical device pages independently, so their page IDs are not guaranteed to match.
This PR carries separate device KV indices through the Mooncake transfer path:

- Host buffers use the physical host page indices.
- C4 indexer and C128 device buffers use the logical device page indices.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #30778103991](https://github.com/sgl-project/sglang/actions/runs/30778103991)
Latest PR Test (Extra): :x: [Run #30778103903](https://github.com/sgl-project/sglang/actions/runs/30778103903)

### CODE REVIEW

**Comment by huangtingwei9988** [comment]
Validation was performed on DeepSeek V4 Flash with 4-GPU prefill and 4-GPU decode.

 GPQA
HiSparse before fix  **81.63%** 
Three representative GPQA runs produced comparable results with and without HiSparse:

| Run | No HiSparse | HiSparse After Fix |
| --- | ---: | ---: |
| 1 | 86.36% | 86.87% |
| 2 | 87.88% | 87.88% |
| 3 | 86.36% | 88.38% |
| **Overall** | **86.87% (516/594)** | **87.71% (521/594)** |

The results confirm that GPQA accuracy returned to the normal range after the fix.

SWE Tool Calls

- 160/160 valid responses across 16 cases and 10 turns.
- 0 malformed tool calls.
- 0 empty tool calls.

**Review by ShangmingCai — APPROVED** [review]
Looks good. But need a rebase to unblock CI.

**Review by ShangmingCai — COMMENTED** [review]
Probably need to check the diff and compatibility with the DCP that was recently added.

**Comment by huangtingwei9988** [comment]
> Probably need to check the diff and compatibility with the DCP that was recently added.

Okay, I'll add support for HiSparse and DCP compatibility in the next PR.
